310 log what course a profile belongs to

matthewhilton opened the pull request:
Closes #310 
Closes #312 

**Changes**

- Bugfix for running unit tests individually ( #312 )

- Logs $COURSE->id when saving profile
- New tab for group by courses (similar to the group by script)
- Drill down page for courses
- Individual profile view shows course

**Screenshots**
Course groups
![image](https://user-images.githubusercontent.com/17095477/236115266-8ea0d76e-5926-495f-ba97-2bcae9ab4a0a.png)

Course drill-down
![image](https://user-images.githubusercontent.com/17095477/236115169-200a24f2-fc25-46a1-a32a-42e26ccf2add.png)

Individual profile
![image](https://user-images.githubusercontent.com/17095477/236115342-4f23ec69-6e26-45f4-ba19-92434b3d7366.png)

matthewhilton committed:
[#312] Fix unit tests when running individually

matthewhilton committed:
[#310] Store course when saving profile

Peterburnett commented on classes/helper.php:
```diff
@@ -17,6 +17,7 @@
 namespace tool_excimer;
 
 use core_filetypes;
+use moodle_url;
```
nitpick: I generally prefer to just \ this inline, purely because its a little cleaner IMO.

Peterburnett commented on classes/output/tabs.php:
```diff
@@ -60,6 +60,12 @@ public function export_for_template(\renderer_base $output): array {
                 ],
+                [
+                    'id' => 'course',
+                    'link' => [['link' => new \moodle_url('/admin/tool/excimer/slow_course.php')]],
```
Why is the link a double array? Im looking at the wider pattern and it seems really bizarre

matthewhilton replied:
Not sure - it seems redundant, since at the bottom it just calls `link[0]->link`

https://github.com/catalyst/moodle-tool_excimer/blob/5257c205ba836337b01b9f5f2780efea73332bcc/classes/output/tabs.php#L90-L94

matthewhilton replied:
Was added in 5a1bad9 - maybe the idea was multiple links per tab?

Peterburnett commented on classes/profile.php:
```diff
@@ -212,6 +212,10 @@ public function save_record(): int {
             $this->raw_set('userid', $USER->id);
         }
 
+        if ($this->check_update_courseid($COURSE->id)) {
```
Is this going to store everything in the admin level under course=1? That might be fine, or it might muddy the waters a little

matthewhilton replied:
Yes correct. I would expect anyone using the Excimer plugin has enough Moodle knowledge to know why course id = 1 is used so often so I don't think it's an issue.

Peterburnett commented on classes/slow_course_table.php:
```diff
@@ -0,0 +1,162 @@
+ * Table for displaying grouped profile lists.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
```
Author needs adjusting

Peterburnett replied:
Also year

Peterburnett commented on classes/slow_course_table.php:
```diff
@@ -0,0 +1,162 @@
+        list($filterstring, $filterparams) = $this->get_filter_for_sql();
+
+        $this->set_count_sql(
+            "SELECT count(distinct request)
```
Just for readability, all the SQL keywords should be caps eg distinct -> DISTINCT as well as the functions in the next query (MAX, COUNT) etc

Peterburnett commented on lang/en/tool_excimer.php:
```diff
@@ -142,6 +144,7 @@
 $string['field_hostname'] = 'Host name';
 $string['field_useragent'] = 'User agent';
 $string['field_versionhash'] = 'Version Hash';
+$string['field_course'] = 'Course';
```
This and 'Courses' I expect would already exist in core, and can be reused

matthewhilton replied:
We can't reuse them since the excimer plugin looks for `get_string('field_' . $fieldname)`

matthewhilton replied:
https://github.com/catalyst/moodle-tool_excimer/blob/5257c205ba836337b01b9f5f2780efea73332bcc/classes/profile_table.php#L90

Peterburnett commented on slow_course.php:
```diff
@@ -0,0 +1,52 @@
+use tool_excimer\profile_table_page;
+use tool_excimer\slow_course_table;
+
+require_once('../../../config.php');
```
Please make this absolute with `__DIR__`, i've seen this have issues before. PHP sometimes screws this up

Peterburnett commented on tests/tool_excimer_helper_test.php:
```diff
@@ -0,0 +1,61 @@
+     */
+    protected function setUp(): void {
+        parent::setUp();
+        $this->resetAfterTest();
```
Can we just put this in the test itself? that way additional tests aren't forced to use it if its not required

Peterburnett commented on tests/tool_excimer_profile_test.php:
```diff
@@ -33,6 +33,8 @@ class tool_excimer_profile_test extends \advanced_testcase {
     protected function setUp(): void {
         parent::setUp();
         $this->resetAfterTest();
+        profile_helper::init();
```
Why are these required. It doesnt look like you are doing anything new in your tests

matthewhilton replied:
Its for #312 - Because otherwise the tests fail because the static vars are not populated yet https://github.com/catalyst/moodle-tool_excimer/blob/5257c205ba836337b01b9f5f2780efea73332bcc/classes/profile_helper.php#L34-L36

I split it into a separate commit 83c3311d0509c562ec3793f73d3b2aef64f5098d but kept it in this Pr since its a tiny change

brendanheywood commented on classes/slow_course_table.php:
```diff
@@ -0,0 +1,162 @@
+     * @param \stdClass $record
+     * @return string
+     */
+    public function col_maxcreated(\stdClass $record): string {
```
can we inherit a lot of these columns from another table class and some some boilerplate?

matthewhilton replied:
Ah probably. I did reuse a lot of code from `grouped_profile_table.php` - inheriting it is probably cleaner. Will update...

matthewhilton committed:
[#310] Make generic grouped profile table

matthewhilton commented:
I have adjusted the code a bit. Now it is structured:

- `grouped_profile_table` is subclasses by `grouped_course_profile_table` and `grouped_script_profile_table`. Each just specifies what column is being grouped and the parent class does the rest.
- Renamed the db column `groupby` -> `scriptgroup` since it was a bit confusing now that we group by more than just the script. Updated all the references to groupby

brendanheywood commented:
Functionally I think this is good to go. UX wise I think there should be an extra drill down between the course tab and the page tab, eg:

1) I start here: https://master.localhost/admin/tool/excimer/slow_course.php
![image](https://github.com/catalyst/moodle-tool_redirects/assets/187449/48b2d06a-01ac-4081-a36f-0119998c3e6a)

2) Then I drill down into a course and see a list of individual profiles:
https://master.localhost/admin/tool/excimer/slow_course.php?courseid=4
![image](https://github.com/catalyst/moodle-tool_redirects/assets/187449/443a9bcf-6315-41d7-99ef-3d8e98685515)

As you can see there is lots of duplicates. I think it would be better if instead it showed the same contents as slowest_web.php but filtered to that course, and then you drill down again to the profiles. 

Maybe this is better as slowest_web.php?courseid=4 or maybes its better inside slow_course.php I don't mind

matthewhilton commented:
Nice idea @brendanheywood - very valid. Probably would need to restructure the pages a bit to accomplish that so probably best for a followup issue.
